contracts/eden/src/elections.cpp

```diff
@@ -0,0 +1,770 @@
+                   eosio::current_time_point() +
+                       eosio::seconds(globals.get().election_round_time_sec +
+                                      globals.get().election_break_time_sec)};
+            }
+            --max_steps;
+         }
+      }
+      state_sing.set(state_variant, contract);
+      return max_steps;
+   }
+
+   current_election_state_active elections::check_active()
+   {
+      eosio::check(state_sing.exists(), "No election is running");
+      auto state = state_sing.get();
+      eosio::check(std::holds_alternative<current_election_state_active>(state),
+                   "The election is not ready for voting");
+      auto result = std::get<current_election_state_active>(state);
+      eosio::check(eosio::current_block_time() < result.round_end,
+                   "Voting period for this round has closed");
+      return result;
+   }
+
+   static eosio::checksum256 adjust_seed(const eosio::checksum256& seed)
+   {
+      char buf[36];
+      memcpy(buf, seed.extract_as_byte_array().data(), 32);
+      eosio::block_timestamp timestamp = eosio::current_block_time();
+      memcpy(buf + 32, &timestamp.slot, 4);
+      return eosio::sha256(buf, 36);
+   }
+
+   struct group_result
+   {
+      eosio::name winner;
+      std::vector<eosio::name> voted;
+      std::vector<eosio::name> missing;
+   };
+
+   template <typename Idx, typename It>
+   static group_result finish_group(current_election_state_post_round& state,
+                                    Idx& group_idx,
+                                    It& iter,
+                                    uint8_t group_size)
+   {
+      // count votes
+      group_result result;
+      std::map<eosio::name, uint8_t> votes_by_candidate;
+      for (uint32_t i = 0; i < group_size; ++i)
+      {
+         if (iter->candidate != eosio::name())
+         {
+            ++votes_by_candidate[iter->candidate];
+            result.voted.push_back(iter->member);
+         }
+         else
+         {
+            result.missing.push_back(iter->member);
+         }
+         iter = group_idx.erase(iter);
+      }
+      auto best = std::max_element(
+          votes_by_candidate.begin(), votes_by_candidate.end(),
+          [](const auto& lhs, const auto& rhs) { return lhs.second < rhs.second; });
+      if (!votes_by_candidate.empty() && 3 * best->second > 2 * group_size)
+      {
+         result.winner = best->first;
+      }
+      return result;
+   }
+
+   std::vector<eosio::name> elections::extract_board()
+   {
+      std::vector<eosio::name> result;
+      auto vote_idx = vote_tb.get_index<"bygroup"_n>();
+      for (auto iter = vote_idx.begin(), end = vote_idx.end(); iter != end;)
+      {
+         result.push_back(iter->member);
+         eosio::check(result.size() < 12, "Too many board members");
+         iter = vote_idx.erase(iter);
+      }
+      eosio::check(!result.empty(), "No board members");
```

Comment:
This will hang the contract if no groups in the previous level reached consensus

Reply:
Yep.  I don't know what we should do in this case.

Reply:
Maybe create an issue to track this